### PR TITLE
refactor and simplify optional imports

### LINKED ISSUE #2266: Improve Exception Handling
## 🚀 Code Quality Improvement

I came across this a few times already:
```python
try:
   # something
except Exception:
   # something
```
It is the worst possible way to handle Exceptions. It is better to catch the specific exception or at least log a message. 

### Alternatives

None. Sooner or later someone has to deal with this anyway :)

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Fixes #2266

refactors code duplication for recurring optional imports like apex, native amp, xla, etc.
it's also backward compatible with e.g. bolts that needs APEX_AVAILABLE etc.

# Before submitting

- [ ] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you to create a separate PR for every change.
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests? 
- [ ] Did you verify new and existing tests pass locally with your changes?
- [ ] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

## PR review    
Anyone in the community is free to review the PR once the tests have passed.     
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by Borda** [comment]
mind update after #2865

**Comment by awaelchli** [comment]
@Borda adding this to 0.9.x milestone just in case I have a typo, I don't want it to be released :)

File: tests/base/models.py [hunk 1]
```diff
@@ -6,15 +6,8 @@
 import torch.nn.functional as F
 from torch.utils.data import DataLoader
 
-from tests.base.datasets import TrialMNIST, AverageDataset, MNIST
-
-try:
-    from test_tube import HyperOptArgumentParser
-except ImportError:
-    # TODO: this should be discussed and moved out of this package
```
Comment (awaelchli) [inline tests/base/models.py:14]: I see it nowhere used here, so it seems we can easily remove it.

File: pytorch_lightning/trainer/data_loading.py [hunk 2]
```diff
@@ -6,44 +6,24 @@
...
 from pytorch_lightning.core import LightningModule
-from pytorch_lightning.utilities import rank_zero_warn
+from pytorch_lightning.utilities import rank_zero_warn, is_xla_available, is_horovod_available
 from pytorch_lightning.utilities.exceptions import MisconfigurationException
 
-try:
-    from torch.utils.data import IterableDataset
-    ITERABLE_DATASET_EXISTS = True
-except ImportError:
-    ITERABLE_DATASET_EXISTS = False
```
Comment (awaelchli) [inline pytorch_lightning/trainer/data_loading.py:20]: removed this, because iterable dataset is part of torch since 1.2 and we support torch >= 1.3

**Pull request closed without merging**